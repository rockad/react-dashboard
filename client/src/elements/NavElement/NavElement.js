import React from 'react';
import {Link as RouterLink, NavLink as RouterNavLink} from 'react-router-dom';
import PropTypes from 'prop-types';

import Button from '@material-ui/core/Button';
import ListItem from '@material-ui/core/ListItem';
import Link from '@material-ui/core/Link';

function isAbsoluteUrl(url) {
    return /^https?:\/\//.test(url);
}

const NavElement = (
    {
        children,
        className,
        to,
        activeClassName,
        nav,
        variant,
        linkVariant,
        ...rest
    },
) => {
    const props = rest;

    if (nav) {
        props.activeClassName = activeClassName || 'active';
    }

    const componentProps = {
        className,
        to,
        ...props,
        component: nav ? RouterNavLink : RouterLink,
    };

    if (variant === 'link' || isAbsoluteUrl(to)) {
        return (
            <Link
                {...componentProps}
                variant={linkVariant}
            >
                {children}
            </Link>
        );
    }

    if (variant === 'list') {
        return (
            <ListItem
                button
                {...componentProps}
            >
                {children}
            </ListItem>
        );
    }

    const buttonProps = {
        ...componentProps,
        variant,
        disableFocusRipple: variant === 'text',
        disableRipple: variant === 'text',
    };

    return (
        <Button
            {...buttonProps}
        >
            {children}
        </Button>
    );
};

NavElement.propTypes = {
    children: PropTypes.node.isRequired,
    to: PropTypes.string.isRequired,
    className: PropTypes.string,
    activeClassName: PropTypes.string,
    nav: PropTypes.bool,
    size: PropTypes.oneOf(['small', 'medium', 'large']),
    variant: PropTypes.oneOf(['text', 'outlined', 'contained', 'fab', 'extendedFab', 'link', 'list']),
    linkVariant: PropTypes.string,
    color: PropTypes.oneOf(['default', 'inherit', 'primary', 'secondary']),
};

NavElement.defaultProps = {
    nav: false,
    variant: 'text',
};

export default NavElement;
