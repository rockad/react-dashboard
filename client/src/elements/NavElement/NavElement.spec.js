import React from 'react';
import NavElement from '.';

describe('Elements', () => {
    describe('NavElement', () => {
        it('should match snapshot', () => {
            const wrapper = shallow(<NavElement to="/">Home</NavElement>);
            expect(wrapper).toMatchSnapshot();
        });

        it('should match snapshot as active menu item', () => {
            const wrapper = shallow(<NavElement nav to="/">Home</NavElement>);
            expect(wrapper).toMatchSnapshot();
        });
    });
});
