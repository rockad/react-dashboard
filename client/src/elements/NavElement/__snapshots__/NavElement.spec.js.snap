// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Elements NavElement should match snapshot 1`] = `
<WithStyles(Button)
  component={[Function]}
  disableFocusRipple={true}
  disableRipple={true}
  to="/"
  variant="text"
>
  Home
</WithStyles(Button)>
`;

exports[`Elements NavElement should match snapshot as active menu item 1`] = `
<WithStyles(Button)
  activeClassName="active"
  component={[Function]}
  disableFocusRipple={true}
  disableRipple={true}
  to="/"
  variant="text"
>
  Home
</WithStyles(Button)>
`;
